File: packages/go/faucet.go
package rdk

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
)

// FaucetOptions configures a faucet request.
type FaucetOptions struct {
	// URL is the faucet endpoint URL. Required.
	URL string
	// Address is the address to fund.
	Address string
	// Denom is the denomination to request. Empty defaults to uqor.
	Denom string
	// HTTP doer (for testing). Nil uses http.DefaultClient.
	HTTP HTTPDoer
}

// FaucetResult is the outcome of a faucet request.
type FaucetResult struct {
	OK     bool
	Status int
	Body   any
}

// RequestFaucet requests testnet funds from a configured faucet URL. The network
// does not publish a fixed faucet endpoint, so a URL must be supplied. It errors
// clearly when no URL is configured rather than guessing one.
func RequestFaucet(ctx context.Context, options FaucetOptions) (FaucetResult, error) {
	if options.URL == "" {
		return FaucetResult{}, errors.New(
			"No faucet URL configured. Set a faucet endpoint (e.g. QORE_FAUCET_URL) or fund the " +
				"account manually — see the keys & funding guide.")
	}
	denom := options.Denom
	if denom == "" {
		denom = "uqor"
	}
	doer := options.HTTP
	if doer == nil {
		doer = http.DefaultClient
	}
	payload, err := json.Marshal(map[string]any{"address": options.Address, "denom": denom})
	if err != nil {
		return FaucetResult{}, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, options.URL, bytes.NewReader(payload))
	if err != nil {
		return FaucetResult{}, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	res, err := doer.Do(req)
	if err != nil {
		return FaucetResult{}, err
	}
	defer res.Body.Close()
	raw, _ := io.ReadAll(res.Body)
	var body any
	_ = json.Unmarshal(raw, &body)
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return FaucetResult{}, fmt.Errorf("Faucet request failed: %d %s", res.StatusCode, http.StatusText(res.StatusCode))
	}
	return FaucetResult{OK: true, Status: res.StatusCode, Body: body}, nil
}

File: packages/rust/src/faucet.rs
//! Testnet faucet helper. The network does not publish a fixed faucet endpoint,
//! so this posts to a URL you supply. It fails with a clear message when no URL
//! is configured rather than guessing one.

use std::collections::HashMap;

use serde_json::{json, Value};
use thiserror::Error;

use crate::client::http::{default_transport, HttpError, HttpRequest, Method, SharedTransport};
use crate::constants::BASE_DENOM;

/// A faucet request error.
#[derive(Debug, Clone, Error)]
pub enum FaucetError {
    /// No faucet URL was configured.
    #[error(
        "No faucet URL configured. Set a faucet endpoint or fund the account manually -- see the keys & funding guide."
    )]
    NoUrl,
    /// The transport failed.
    #[error(transparent)]
    Transport(#[from] HttpError),
    /// The faucet returned a non-2xx status.
    #[error("Faucet request failed: {0}")]
    Status(u16),
}

/// The result of a faucet request.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FaucetResult {
    /// The HTTP status code.
    pub status: u16,
    /// The parsed response body, if any.
    pub body: Option<Value>,
}

/// Request testnet funds from a configured faucet URL.
pub fn request_faucet(
    url: Option<&str>,
    address: &str,
    denom: Option<&str>,
    transport: Option<SharedTransport>,
) -> Result<FaucetResult, FaucetError> {
    let url = match url {
        Some(u) if !u.trim().is_empty() => u,
        _ => return Err(FaucetError::NoUrl),
    };
    let transport = transport.unwrap_or_else(default_transport);
    let mut headers = HashMap::new();
    headers.insert("content-type".to_string(), "application/json".to_string());
    headers.insert("accept".to_string(), "application/json".to_string());
    let body = json!({ "address": address, "denom": denom.unwrap_or(BASE_DENOM) });
    let resp = transport.send(HttpRequest {
        method: Method::Post,
        url: url.to_string(),
        headers,
        body: Some(body.to_string()),
    })?;
    if !resp.ok() {
        return Err(FaucetError::Status(resp.status));
    }
    let parsed = serde_json::from_str(&resp.body).ok();
    Ok(FaucetResult {
        status: resp.status,
        body: parsed,
    })
}
